File: experiments/bakeoff/benchmarks/dart-inventory/ground_truth/reference/lib/persistence.dart
/// persistence.dart — In-memory store shared by all services.
///
/// Holds maps keyed by id/sku/orderId/tokenId. `snapshot()` returns a
/// deep-enough copy for round-trip restore in tests. `restore()`
/// replaces all current state with a snapshot.

import 'types/customers.dart';
import 'types/products.dart';
import 'types/inventory.dart';
import 'types/orders.dart';

class Snapshot {
  final Map<String, Customer> customers;
  final Map<String, Product> products;
  final Map<String, StockLevel> stock;
  final Map<String, Order> orders;
  final Map<String, ReservationToken> reservations;

  Snapshot({
    required this.customers,
    required this.products,
    required this.stock,
    required this.orders,
    required this.reservations,
  });
}

class Store {
  final Map<String, Customer> customers = <String, Customer>{};
  final Map<String, Product> products = <String, Product>{};
  final Map<String, StockLevel> stock = <String, StockLevel>{};
  final Map<String, Order> orders = <String, Order>{};
  final Map<String, ReservationToken> reservations =
      <String, ReservationToken>{};

  Snapshot snapshot() {
    return Snapshot(
      customers: Map<String, Customer>.from(customers),
      products: Map<String, Product>.from(products),
      stock: stock.map(
        (k, v) => MapEntry(
          k,
          StockLevel(sku: v.sku, onHand: v.onHand, reserved: v.reserved),
        ),
      ),
      orders: orders.map(
        (k, v) => MapEntry(
          k,
          Order(
            id: v.id,
            customerId: v.customerId,
            items: List<Item>.from(v.items),
            status: v.status,
            history: List<Transition>.from(v.history),
            reservationTokens: List<String>.from(v.reservationTokens),
          ),
        ),
      ),
      reservations: reservations.map(
        (k, v) => MapEntry(
          k,
          ReservationToken(
            tokenId: v.tokenId,
            orderId: v.orderId,
            sku: v.sku,
            quantity: v.quantity,
            committed: v.committed,
            released: v.released,
          ),
        ),
      ),
    );
  }

  void restore(Snapshot snap) {
    customers
      ..clear()
      ..addAll(snap.customers);
    products
      ..clear()
      ..addAll(snap.products);
    stock
      ..clear()
      ..addAll(snap.stock);
    orders
      ..clear()
      ..addAll(snap.orders);
    reservations
      ..clear()
      ..addAll(snap.reservations);
  }
}

File: experiments/bakeoff/benchmarks/python-inventory/ground_truth/reference/shop/persistence.py
"""persistence.py — In-memory store shared by all services.

Holds dicts keyed by id/sku/order_id/token_id. `snapshot()` returns a
deep-enough copy for round-trip restore in tests. `restore()`
replaces all current state with a snapshot.
"""

from copy import deepcopy
from dataclasses import dataclass, field
from typing import Dict

from .types.customers import Customer
from .types.inventory import ReservationToken, StockLevel
from .types.orders import Order
from .types.products import Product


@dataclass
class Snapshot:
    customers: Dict[str, Customer] = field(default_factory=dict)
    products: Dict[str, Product] = field(default_factory=dict)
    stock: Dict[str, StockLevel] = field(default_factory=dict)
    orders: Dict[str, Order] = field(default_factory=dict)
    reservations: Dict[str, ReservationToken] = field(default_factory=dict)


@dataclass
class Store:
    customers: Dict[str, Customer] = field(default_factory=dict)
    products: Dict[str, Product] = field(default_factory=dict)
    stock: Dict[str, StockLevel] = field(default_factory=dict)
    orders: Dict[str, Order] = field(default_factory=dict)
    reservations: Dict[str, ReservationToken] = field(default_factory=dict)

    def snapshot(self) -> Snapshot:
        return Snapshot(
            customers=dict(self.customers),
            products=dict(self.products),
            stock={k: deepcopy(v) for k, v in self.stock.items()},
            orders={k: deepcopy(v) for k, v in self.orders.items()},
            reservations={k: deepcopy(v) for k, v in self.reservations.items()},
        )

    def restore(self, snap: Snapshot) -> None:
        self.customers.clear()
        self.customers.update(snap.customers)
        self.products.clear()
        self.products.update(snap.products)
        self.stock.clear()
        self.stock.update(snap.stock)
        self.orders.clear()
        self.orders.update(snap.orders)
        self.reservations.clear()
        self.reservations.update(snap.reservations)
